Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -121,21 +123,24 @@ void WebPermissionController::tryProcessingRequests()
     }
 }
 
-void WebPermissionController::addObserver(WebCore::PermissionObserver& observer)
+void WebPermissionController::addObserver(WebCore::PermissionStatusIdentifier statusIdentifier, WebCore::ClientOrigin&& origin, WebCore::PermissionDescriptor descriptor, WebCore::ScriptExecutionContextIdentifier contextIdentifier)
 {
-    m_observers.add(observer);
+    ASSERT(isMainRunLoop());
+    m_observers.add(statusIdentifier, ObserverData { WTFMove(origin), descriptor, contextIdentifier });
 }
 
-void WebPermissionController::removeObserver(WebCore::PermissionObserver& observer)
+void WebPermissionController::removeObserver(WebCore::PermissionStatusIdentifier statusIdentifier)
 {
-    m_observers.remove(observer);
+    ASSERT(isMainRunLoop());
+    m_observers.remove(statusIdentifier);
 }
 
 void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& origin, const WebCore::PermissionDescriptor& descriptor, WebCore::PermissionState state)
 {
-    for (auto& observer : m_observers) {
-        if (observer.origin() == origin && observer.descriptor() == descriptor)
-            observer.stateChanged(state);
+    ASSERT(isMainRunLoop());
+    for (auto& [permissionStatusIdentifier, observerData] : m_observers) {
+        if (observerData.origin == origin && observerData.descriptor == descriptor)
```

Comment:
It would be nice to reverse the checks since checking descriptor equality is cheaper.

Reply:
Fixed!